sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -1043,6 +1044,22 @@ object TypeCoercion {
       }
     }
   }
+
+  /**
+   * A special rule to support string literal as the second argument of date_add/date_sub functions,
+   * to keep backward compatibility.
+   * TODO: revisit the type coercion rules for string.
+   */
+  object StringLiteralCoercion extends TypeCoercionRule {
+    override protected def coerceTypes(plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
+      // Skip nodes who's children have not been resolved yet.
+      case e if !e.childrenResolved => e
+      case DateAdd(l, r) if r.dataType == StringType && r.foldable =>
+        DateAdd(l, AnsiCast(r, IntegerType))
```

Comment:
We need to depend on `AnsiCast` here, even though we can know cast errors in an analysis phase? IMHO an analysis exception with an explicit error message is better than the current number format exception: https://github.com/apache/spark/pull/27965/files#diff-79dd276be45ede6f34e24ad7005b0a7cR279

Reply:
Also, is the behaviour independent of the ANSI mode? Any string->int casts should not be allowed when `ansi=true`.

Reply:
> Any string->int casts should not be allowed when ansi=true.

Do you mean type coercion? Many databases can cast string to int and they are ANSI-compliant.

Reply:
Ah, yes. I rechecked the reference of implicit casts in some databases, as you said, the case seems to be acceptable.